### PR TITLE
Add multi-context seeds

### PR DESCRIPTION
**Pull request by Itolstoganov**
This replaces randstrobe hashes with multi-context hashes.  Multi-context hashes are defined as follows

`((hash1 >> digest_size) << digest_size) ^ (hash2 >> (64 - digest_size))`

Where `hash1` and `hash2` are the syncmer hashes, with `hash1` being the smaller hash. The `64 - digest_size` prefix of the multi-context hash corresponding to `hash1` is denoted as the _main_ hash, and the `digest_size` suffix corresponding to `hash2` is denoted as _auxiliary_ hash.

The prefix of the size (`64 - digest_size`) corresponding to the smaller syncmer is used for partial search. If a given full 64-bit query hash was not found in the index, the seeds with the same main hash are looked up instead and added as the matches. These _partial_ matches are used for the NAM construction in the same way as the full matches.

In order to calculate the reference range of the partial match correctly, we need to know which strobe was used as main. This information is stored in the `m_packed` field of `RefRandstrobe`, replacing one of the bits previously reserved for the reference index.

List of changes:

- The randstrobe hash was replaced with the multi-context hash.
- Partial match search was added to the `find_nams` function if full matches were not found. Partial matches have a larger abundance threshold than full matches due to higher seed repetitiveness.
- Partial hit processing added to `add_to_hits_per_ref`.
- The `partial_find` function that looks for the seeds with the same main hash as the query was added to the `StrobemerIndex`. 
- Some utility functions (for working with the main part of the hash) were added to the `StrobemerIndex`.
- 9th right-hand bit of the `m_packed` field of the `RefRandstrobe` is now set iff the first strobe was used as the main part of the multi-context seed.
- Fields that store the start and end positions of the main strobe on the query, and the indicator of the main strobe are added to the `QueryRandstrobe` class.
- The `--digest` parameter was added to regulate the auxiliary part of the multi-context hash with the default value of 24.

### CODE REVIEW

File: src/randstrobes.cpp [hunk 1]
```diff
@@ -49,6 +49,12 @@ static inline randstrobe_hash_t randstrobe_hash(syncmer_hash_t hash1, syncmer_ha
     return hash1 + hash2;
 }
 
+static inline digest_hash_t digest_hash(syncmer_hash_t hash1, syncmer_hash_t hash2, size_t digest_size) {
+    syncmer_hash_t main_hash = std::min(hash1, hash2);
+    syncmer_hash_t aux_hash = std::max(hash1, hash2);
+    return ((main_hash >> digest_size) << digest_size) ^ (aux_hash >> (64 - digest_size));
```
Comment (ksahlin) [inline src/randstrobes.cpp:55]: Need to pass information of which strobe (1 or 2) that was saved as the base value. probably here

File: src/nam.cpp [hunk 2]
```diff
@@ -14,12 +15,14 @@ inline void add_to_hits_per_ref(
     int query_start,
     int query_end,
     const StrobemerIndex& index,
-    size_t position
+    size_t position,
+    bool is_partial
 ) {
     int min_diff = std::numeric_limits<int>::max();
     for (const auto hash = index.get_hash(position); index.get_hash(position) == hash; ++position) {
         int ref_start = index.get_strobe1_position(position);
```
Comment (ksahlin) [inline src/nam.cpp:23]: if using multi-context, we are not sure if it is index.get_strobe1_position(position) or position of the second strobe here. Will yield inconsistent NAMs

File: src/randstrobes.cpp (hunk 1 above)
Comment (Itolstoganov) [reply src/randstrobes.cpp]: Added this information to the last bit of the hash

File: src/nam.cpp (hunk 2 above)
Comment (Itolstoganov) [reply src/nam.cpp]: Fixed

**Review by marcelm — COMMENTED** [review]
> This replaces randstrobes with multi-context seeds.

For practical reasons, I’d like to argue that we still have randstrobes, we just compute their hash differently. If we didn’t have randstrobes anymore, we’d have to rename *a lot* of functions, files and variables. So I see it as "replaces randstrobe hashes with multi-context hashes" or something like that.

> Where `hash1` and `hash2` are the syncmer hashes, with `hash1` being the smaller hash.

Is this so that randstrobe hashes are symmetric as before? That is, would that be different when we switch to asymmetric hashes?

> The `64 - digest_size` prefix of the multi-context hash corresponding to `hash1` is denoted as the _main_ hash, and the `digest_size` suffix corresponding to `hash2` is denoted as _auxiliary_ hash.

This PR introduces quite a bit of terminology that is sometimes overlapping and sometimes used inconsistently, and I wonder whether that could be simplified a bit.

* main hash
* auxiliary hash
* digest, digest size (inconsistency: command-line option is `--digest`, but what is meant is actually "digest length"). What is a digest anyway?
* subhash (used in a comment somewhere)
* multi-context hash
* digest hash
* partial hit

I guess that some of this will resolve itself if we switch to asymmetric randstrobe hashes, so it is not super important to fix this at the moment.

File: src/index.cpp [hunk 3]
```diff
@@ -150,6 +150,8 @@ void StrobemerIndex::populate(float f, unsigned n_threads) {
     }
     stats.tot_strobemer_count = total_randstrobes;
 
+    logger.info() << "Digest parameter is : " << parameters.randstrobe.digest;
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/index.cpp]: Replaced "digest" with "aux_len" here and in other places.

File: src/dumpstrobes.cpp [hunk 4]
```diff
@@ -125,7 +125,8 @@ int run_dumpstrobes(int argc, char **argv) {
         l_set ? l : IndexParameters::DEFAULT,
         u_set ? u : IndexParameters::DEFAULT,
         c_set ? c : IndexParameters::DEFAULT,
-        max_seed_len_set ? max_seed_len : IndexParameters::DEFAULT
+        max_seed_len_set ? max_seed_len : IndexParameters::DEFAULT,
+        digest ? digest: IndexParameters::DEFAULT
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/dumpstrobes.cpp]: Fixed

File: src/index.hpp [hunk 5]
```diff
@@ -80,18 +83,66 @@ struct StrobemerIndex {
         return end();
     }
 
+    //Returns the first entry that matches the first strobe subhash (if using multi-context seeds)
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/index.hpp]: It is not optional, but I wanted to leave the possibility of changing between hashes open at some point. But it does not seem to have any benefits. Fixed the comment.

File: src/indexparameters.cpp [hunk 6]
```diff
@@ -30,25 +31,26 @@ struct Profile {
...
-        Profile{150, 175, 20, -4,  1,  7},
-        Profile{250, 375, 22, -4,  2, 12},
-        Profile{400, max, 23, -6,  2, 12},
+        Profile{ 50,  70, 18, -4, -2,  1, 24},
+        Profile{ 75,  90, 20, -4, -3,  2, 24},
+        Profile{100, 110, 20, -4, -2,  2, 24},
+        Profile{125, 135, 20, -4, -1,  4, 24},
+        Profile{150, 175, 20, -4,  1,  7, 24},
+        Profile{250, 375, 22, -4,  2, 12, 24},
+        Profile{400, max, 23, -6,  2, 12, 24},
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/indexparameters.cpp]: Removed

File: src/nam.cpp [hunk 7]
```diff
@@ -14,16 +15,43 @@ inline void add_to_hits_per_ref(
...
 ) {
     int min_diff = std::numeric_limits<int>::max();
-    for (const auto hash = index.get_hash(position); index.get_hash(position) == hash; ++position) {
-        int ref_start = index.get_strobe1_position(position);
-        int ref_end = ref_start + index.strobe2_offset(position) + index.k();
-        int diff = std::abs((query_end - query_start) - (ref_end - ref_start));
-        if (diff <= min_diff) {
-            hits_per_ref[index.reference_index(position)].push_back(Hit{query_start, query_end, ref_start, ref_end});
-            min_diff = diff;
+    if (not is_partial) {
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/nam.cpp]: Good point! Splitted the function.

File: src/randstrobes.hpp [hunk 8]
```diff
@@ -16,6 +16,7 @@
 
 using syncmer_hash_t = uint64_t;
 using randstrobe_hash_t = uint64_t;
+using digest_hash_t = uint64_t;
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/randstrobes.hpp]: Removed

File: src/index.hpp [hunk 9]
```diff
@@ -80,18 +83,66 @@ struct StrobemerIndex {
...
+    bool is_first_strobe_main(bucket_index_t position) const {
+        return randstrobes[position].is_first_main();
+    }
     
     bool is_filtered(bucket_index_t position) const {
         return get_hash(position) == get_hash(position + filter_cutoff);
     }
 
+    bool is_partial_filtered(bucket_index_t position) const {
+        uint shift = parameters.randstrobe.digest;
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/index.hpp]: Replaced all uint occurences

File: src/randstrobes.cpp [hunk 10]
```diff
@@ -49,6 +49,13 @@ static inline randstrobe_hash_t randstrobe_hash(syncmer_hash_t hash1, syncmer_ha
     return hash1 + hash2;
 }
 
+static inline digest_hash_t digest_hash(syncmer_hash_t hash1, syncmer_hash_t hash2, size_t digest_size) {
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/randstrobes.cpp]: Fixed

**Comment by Itolstoganov** [comment]
> For practical reasons, I’d like to argue that we still have randstrobes, we just compute their hash differently. If we didn’t have randstrobes anymore, we’d have to rename a lot of functions, files and variables. So I see it as "replaces randstrobe hashes with multi-context hashes" or something like that.

Sure, edited the PR description.

> Where hash1 and hash2 are the syncmer hashes, with hash1 being the smaller hash.
> Is this so that randstrobe hashes are symmetric as before? That is, would that be different when we switch to asymmetric hashes?

Yes, the only reason for selecting minimal hash is to keep the hash symmetric.

> This PR introduces quite a bit of terminology that is sometimes overlapping and sometimes used inconsistently, and I wonder whether that could be simplified a bit.

I tried to alleviate that by removing all mentions of "digest" (by which I meant the same thing as the "auxiliary" part of the multi-context hash) and "subhash".

File: src/nam.cpp [hunk 11]
```diff
@@ -231,12 +272,14 @@ std::vector<Nam> find_nams_rescue(
...
             }
-            add_to_hits_per_ref(hits_per_ref[is_revcomp], rh.query_start, rh.query_end, index, rh.position);
+            add_to_hits_per_ref(hits_per_ref[is_revcomp], rh.query_start, rh.query_end, index, rh.position, false);
             cnt++;
         }
         is_revcomp++;
     }
 
+
+
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/nam.cpp]: Removed

File: src/randstrobes.cpp [hunk 12]
```diff
@@ -167,7 +175,12 @@ Randstrobe RandstrobeIterator::get(unsigned int strobe1_index) const {
         }
     }
 
-    return Randstrobe{randstrobe_hash(strobe1.hash, strobe2.hash), static_cast<uint32_t>(strobe1.position), static_cast<uint32_t>(strobe2.position)};
+//    return Randstrobe{randstrobe_hash(strobe1.hash, strobe2.hash), static_cast<uint32_t>(strobe1.position), static_cast<uint32_t>(strobe2.position)};
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/randstrobes.cpp]: Removed

File: src/randstrobes.cpp [hunk 13]
```diff
@@ -49,6 +49,13 @@ static inline randstrobe_hash_t randstrobe_hash(syncmer_hash_t hash1, syncmer_ha
     return hash1 + hash2;
 }
 
+static inline digest_hash_t digest_hash(syncmer_hash_t hash1, syncmer_hash_t hash2, size_t digest_size) {
+    if (hash1 < hash2) {
+        return ((hash1 >> digest_size) << digest_size) ^ (hash2 >> (64 - digest_size));
+    }
+    return ((hash2 >> digest_size) << digest_size) ^ (hash1 >> (64 - digest_size));
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/randstrobes.cpp]: Fixed, thanks!

File: src/randstrobes.cpp [hunk 14]
```diff
@@ -198,7 +211,12 @@ Randstrobe RandstrobeGenerator::next() {
         }
     }
     syncmers.pop_front();
-    return Randstrobe{randstrobe_hash(strobe1.hash, strobe2.hash), static_cast<uint32_t>(strobe1.position), static_cast<uint32_t>(strobe2.position)};
+//    return Randstrobe{randstrobe_hash(strobe1.hash, strobe2.hash), static_cast<uint32_t>(strobe1.position), static_cast<uint32_t>(strobe2.position)};
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/randstrobes.cpp]: Removed

File: src/randstrobes.hpp [hunk 15]
```diff
@@ -39,8 +40,12 @@ struct RefRandstrobe {
         return lhs < rhs;
     }
 
+    bool is_first_main() const {
```
[the remark this answers is not in the document]
Comment (Itolstoganov) [reply src/randstrobes.hpp]: I am used to having "is_" as the prefix, but `main_is_first` looks more intuitive, replaced.

File: src/index.cpp [hunk 16]
```diff
@@ -150,6 +150,8 @@ void StrobemerIndex::populate(float f, unsigned n_threads) {
     }
     stats.tot_strobemer_count = total_randstrobes;
 
+    logger.info() << "Auxiliary hash length is : " << parameters.randstrobe.aux_len;
```
Comment (marcelm) [inline src/index.cpp:153]:
```suggestion
    logger.info() << "Auxiliary hash length: " << parameters.randstrobe.aux_len;
```

**Comment by ksahlin** [comment]
Is this PR ready for evaluation?

**Comment by ksahlin** [comment]
The randstrobe iterator for queries stops when there is still `w_min` syncmers left in the read by checking 
````
    bool has_next() {
        return strobe1_index + w_min < syncmers.size();
````

This is expected behaviour for our current seeds. For example, `w_min=1` for 50 and 75, and `w_min=4` for 100. 

However, I think mcs can be boosted further by adding the remaining syncmers 'in the ends of reads' as seeds. This means 2*w_min more seeds for a read (fw and rc ends). 

I haven't thought carefully about the best way to change it in the code, but perhaps changing the `bool has_next()` to `return strobe1_index < syncmers.size();` together with adding a case to return randstrobes with base/main hash as `strobe1.hash` and auxillary hash 0 in `Randstrobe RandstrobeIterator` when `i < w_start`?

**Comment by ksahlin** [comment]
I implemented and tested this briefly:

1. Changed to 
```
    bool has_next() {
        return strobe1_index < syncmers.size();
    }
```

2. Added in `RandstrobeIterator::get`
```
    if (syncmers.size() < w_start) {
        return Randstrobe{
                randstrobe_hash(strobe1.hash, strobe2.hash, aux_len),
                static_cast<uint32_t>(strobe1.position),
                static_cast<uint32_t>(strobe2.position), true
        };
    }
```

The results only very slightly improved/nearly unchanged on a 'no variation' drosophila genome in SE mapping mode, but they improve substantially for a high variation simulation (see below for read lengths 100 and 150). This also suggest that we may be overfitting our optimization to high quality simulations without much variation.

```
ref,read_length,tool,%-aligned,accuracy,time(sec),Mem(MB)
droso_above10k_variation,100,strobealign-mcs-fix_SE,align,48.815,44.2065,704,1.45,778.4
droso_above10k_variation,100,strobealign-mcs-more-seeds_SE,align,48.9885,44.2975,704,1.55,778.36

droso_above10k_variation,150,strobealign-mcs_SE,align,49.0685,44.9805,839,2.47,781.87
droso_above10k_variation,150,strobealign-mcs-more-seeds_SE,align,49.258,45.1925,839,2.8,787.6

**Comment by marcelm** [comment]
> This also suggest that we may be overfitting our optimization to high quality simulations without much variation.

I am a bit worried about this. Doesn’t this fit the pattern that we often see worse variant detection rates for "optimized" mapping parameters? Shall I perhaps run the parameter optimization on data with higher variation?

**Comment by ksahlin** [comment]
> Shall I perhaps run the parameter optimization on data with higher variation?

Sounds like a good suggestion to me! From what you told me, it should come with relatively little extra computation time since the same index can be used for read sets with different levels of variation w.r.t. the reference genome.

Below were the variant rates I used for SIM1-4. But I think SIM4 is the only setting that really tests aligning arouind variants properly. I would maybe even set `--sv-indel-rate 0.00002 --snp-rate 0.005 --small-indel-rate 0.001 --max-small-indel-size 100` to make it really challenging.

```
rule mason_simulate_variants:
    input:  ref = config["HG38"]
    output: sim_vcf =  config["ROOT_OUT"] + "/reads/{dataset}/variations.vcf",
    run:
        if wildcards.dataset == "SIM1":
            shell("mason_variator -ir {input.ref} -ov {output.sim_vcf}")
        elif wildcards.dataset == "SIM2":
            shell("mason_variator --sv-indel-rate 0.000001 --snp-rate 0.001 --small-indel-rate 0.00001 --max-small-indel-size 20  -ir {input.ref} -ov {output.sim_vcf}")
        elif wildcards.dataset == "SIM3":
            shell("mason_variator --sv-indel-rate 0.000005 --snp-rate 0.001 --small-indel-rate 0.0001 --max-small-indel-size 50   -ir {input.ref} -ov {output.sim_vcf}")
        elif wildcards.dataset == "SIM4":
            shell("mason_variator --sv-indel-rate 0.00001 --snp-rate 0.005 --small-indel-rate 0.0005 --max-small-indel-size 50   -ir {input.ref} -ov {output.sim_vcf}")


```

**Comment by marcelm** [comment]
> From what you told me, it should come with relatively little extra computation time since the same index can be used for read sets with different levels of variation w.r.t. the reference genome.

Sorry if you got the wrong impression, but since I don’t store the index on disk, everything has to be recomputed. I’ll run this on a cluster somewhere to get the results faster this time.

> Below were the variant rates I used for SIM1-4. But I think SIM4 is the only setting that really tests aligning arouind variants properly. I would maybe even set `--sv-indel-rate 0.00002 --snp-rate 0.005 --small-indel-rate 0.001 --max-small-indel-size 100` to make it really challenging.

Good, I’ll use that and would like to call it `SIM5`; is that ok?

**Comment by ksahlin** [comment]
> Good, I’ll use that and would like to call it SIM5; is that ok?

Yes, sounds good!

**Comment by marcelm** [comment]
This PR has been reworked and became #426, which now supersedes it. I’ve read through the comments here and think we have addressed all relevant ones, so I’m closing this PR.

**Pull request closed without merging**